core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -820,24 +821,64 @@ void finishBecauseUnavailable(ShardId shardId, String message) {
         }
     }
 
+
+    static ConcurrentMap<IndexShardReference, String> openShardReferences;
```

Comment:
guys we can't have static maps on classes like this. If we want some kind of static assertions we should inject some mock services just as we do for for `SearchContext` in `MockSearchService`.  I am a bit confused why we have a static factory method on literally the only impl of an interface in  `IndexShardReferenceImpl` this smells like a design flaw to me. I think with this change together we should rather expose a pluggalbe service or move the creation into `IndexService` and let folks plug in a factory into `IndexModule` in order to impl this assertion? but static per JVM concurrent maps seems like a playing russian roulette